src/main/java/pro/belbix/ethparser/web3/harvest/HardWorkCalculator.java

```diff
@@ -0,0 +1,132 @@
+package pro.belbix.ethparser.web3.harvest;
+
+import org.springframework.stereotype.Service;
+
+import pro.belbix.ethparser.dto.v0.HardWorkDTO;
+import pro.belbix.ethparser.dto.v0.HarvestDTO;
+import pro.belbix.ethparser.repositories.v0.HardWorkRepository;
+import pro.belbix.ethparser.repositories.v0.HarvestRepository;
+import pro.belbix.ethparser.web3.EthBlockService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class HardWorkCalculator {
+
+    private final double ETH_ESTIMATE = 0.1;
+    private final HarvestRepository harvestRepository;
+    private final HardWorkRepository hardworkRepository;
+    private final EthBlockService ethBlockService;
+
+    public HardWorkCalculator(HarvestRepository harvestRepository, HardWorkRepository hardWorkRepository, EthBlockService ethBlockService) {
+        this.harvestRepository = harvestRepository;
+        this.hardworkRepository = hardWorkRepository;
+        this.ethBlockService = ethBlockService;
+    }
+
+    public double calculateTotalHardWorksFeeByOwner(String ownerAddress) {
+        long lastBlock = ethBlockService.getLastBlock();
+        long lastBlockDate = ethBlockService.getTimestampSecForBlock(null, lastBlock);
+        
+        HashMap<String, ArrayList<long[]>> blockRangeByVault = new HashMap<>();
+        List<HarvestDTO> harvests = harvestRepository.fetchAllByOwner(ownerAddress, 0, lastBlockDate);
+        
+        harvests
+            .stream()
+            .filter(harvest -> !harvest.getVault().startsWith("PS"))
+            .forEach(harvest -> {
+                String vault = harvest.getVault();
+                double balance = harvest.getOwnerBalance();
+    
+                if (!blockRangeByVault.containsKey(vault)) {
+                    if (balance == 0) {
+                        return;
+                    }
+                    ArrayList<long[]> blockPeriods = new ArrayList<>();
+                    blockRangeByVault.put(vault, blockPeriods);
+                }
+    
+                updateBlockPeriods(blockRangeByVault.get(vault), harvest);
+            });
+
+        return blockRangeByVault
+            .entrySet()
+            .stream()
+            .map(entry -> {
+                String vault = entry.getKey();
+
+                ArrayList<long[]> blockPeriods = entry.getValue();
+                long[] period = calculatePeriod(blockPeriods, lastBlockDate);
+
+                List<HardWorkDTO> allHardWorksByVaultAndPeriod = hardworkRepository.findAllByVaultOrderByBlockDate(vault, period[0], period[1]);
```

Comment:
use (vault = in vaults) and minPeriod + maxPeriod that you found in all vaults
I saw your point
you want to reduce memory usage
but we have 3279 doHardWork records for 4 months, it is not a big amount
if a user will have 20 vaults and request ~100ms you will straggle the HTTP request + DB connection for 2sec
all this time you will keep in memory a ton of other objects
reducing requests quantity is the highest priority

Reply:
it is not a perfect solution in both cases, just for current specific better to reduce queries quantity

Reply:
> but we have 3279 doHardWork records for 4 months, it is not a big amount

But it would grow for the time being, so in theory, the _one go_ query would get slower as time goes.

> if a user will have 20 vaults and request ~100ms you will straggle the HTTP request + DB connection for 2sec

How about running the queries in parallel to avoid the accumulated latency? 

The other possible strategies would be optimizing the network routing between the API and database in order to achieve < 10ms latency or sharding the database for handling the potential heavy concurrent calls in the future. These ways would help maintain the complexity in the code while encapsulating most of the performance optimization via DevOps.